certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -616,26 +596,48 @@ def _enable_redirect(self, domain, unused_options):
         vhost = None
         # If there are blocks listening plaintextishly on self.DEFAULT_LISTEN_PORT,
         # choose the most name-matching one.
+
+        # TODO: should we be adding redirects to all matching blocks?
+
         vhost = self.choose_redirect_vhost(domain, port)
 
         if vhost is None:
             logger.info("No matching insecure server blocks listening on port %s found.",
                 self.DEFAULT_LISTEN_PORT)
+            return
+
+        if vhost.ssl:
+            new_vhost = self.parser.duplicate_vhost(vhost,
+                only_directives=['listen', 'server_name'])
+
+            def _ssl_match_func(directive):
+                return 'ssl' in directive
+
+            def _no_ssl_match_func(directive):
+                return 'ssl' not in directive
+
+            # remove all ssl addresses from the new block
+            self.parser.remove_server_directives(new_vhost, 'listen', match_func=_ssl_match_func)
+
+            # remove all non-ssl addresses from the existing block
+            self.parser.remove_server_directives(vhost, 'listen', match_func=_no_ssl_match_func)
```

Comment:
If it's not the behavior we want, then the way to solve it (and also solve it orthogonally to the other changes here) is to add `default_server` to the `listen 443 ssl` if it's on the `listen 80`. But we haven't been doing that so far, so I don't think it's necessary to do here. If we want to separately make that change then we can, and it will just automatically be taken up by the rest of the code here.

Reply:
> If it's not the behavior we want, then the way to solve it (and also solve it orthogonally to the other changes here) is to add default_server to the listen 443 ssl if it's on the listen 80. 

How does changing the default server block for a different port help? The thing to potentially worry about is the server block serving requests for domains not specified in `server_name` directives suddenly changing. My understanding from [this page](https://nginx.org/en/docs/http/request_processing.html) is in the case there isn't an explicitly specified default server for a domain, nginx chooses the first server block it parsed for that port. By removing this, traffic that may have been being served by one server block may now be served by a completely different one. This seems wrong.

(With that said, the new server block we're creating redirects all traffic to HTTPS. In the case where the domain we're looking at is specified in the default server block, should we change the default behavior of their server to redirect to HTTPS or should it continue to serve HTTP? While more complicated to configure, the latter seems more correct. It's important to note though that we have this problem with our conditional redirects in master, no one has complained about it, and this should almost certainly not be changed in the same PR as this one, but in thinking about this, I feel like our current approach also isn't quite right regarding default servers.)

Reply:
Let's say they have:

```
server {
  server_name example.org;
  listen 80 default_server;
}
```
Right now we would do:
```
server {
  server_name example.org;
  listen 80 default_server;
  listen 443 ssl; # note no default_server is added here, so it's 
                  # not explicitly made the default server for 443
  # rest of the ssl directives
  if (scheme) {
    redirect http->https;
  }
}
```
Instead, we will be doing:
```
server {
  server_name example.org;
  listen 80 default_server;
  redirect http->https;
}
server {
  server_name example.org;
  listen 443 ssl; # we're just as implicitly the default_server as we were before.
  # rest of the ssl directives
}
```
This is the same exact behavior as we had before.

Reply:
Per my understanding of the code and testing locally, the new result is:
```
server {
  server_name example.org;
  listen 80; # notice default_server is removed
  redirect http->https;
}
server {
  server_name example.org;
  listen 443 ssl; 
  # rest of the ssl directives
}
```
I'm not talking about `default_server` on the 443 `listen` statement. Instead, I'm talking about the one for port 80. Now the default server for port 80 is whichever one nginx parses first because it's not specified explicitly. It may be the same block, but it certainly may not. This is a change in behavior.

Reply:
Oh! That's a bug on my part.

Currently, the `duplicate_vhost` method has the parameter `delete_default=True`, under the assumption that if you're duplicating a server you shouldn't be making a second one with `default_server` in there. But given this thing that happened just now, from the person who literally just wrote the other code, I think I'm going to change the default to `False`.